cg/apps/lims/api.py

```diff
@@ -85,74 +85,114 @@ def _export_artifact(self, lims_artifact):
+                if artifact.parent_process and artifact.parent_process.udf.get(udf_key):
+                    delivered_dates.append((artifact.parent_process.date_run,
+                                            artifact.parent_process.udf.get(udf_key)))
+
+        sorted_dates = sorted(delivered_dates, key=lambda x: x[0])
+
+        if sorted_dates:
             log.warning(f"multiple delivery artifacts found for: {lims_id}")
-            return min(artifact.parent_process.udf['Date delivered'] for artifact in artifacts)
+
+        return sorted_dates[-1][1] if sorted_dates else None
 
     def get_sequenced_date(self, lims_id: str) -> dt.date:
         """Get the date when a sample was sequenced."""
-        process_types = ['CG002 - Illumina Sequencing (Illumina SBS)',
-                         'CG002 - Illumina Sequencing (HiSeq X)']
-
-        for process_type in process_types:
-            lims_artifacts = self.get_artifacts(process_type=process_type, samplelimsid=lims_id)
-
-            if len(lims_artifacts) == 0:
-                continue  # continue in case the artifacts are in the other process_type
-            elif len(lims_artifacts) == 1:
-                return lims_artifacts[0].parent_process.udf.get('Finish Date')  # return date immediately if only one artifact is found
-            else:
-                log.warning(f"multiple sequence artifacts found for: {lims_id}")
-                sequence_dates = [artifact.parent_process.udf.get('Finish Date') for artifact in lims_artifacts
-                                  if artifact.parent_process.udf.get('Finish Date') is not None]
-                if len(sequence_dates) == 0:
-                    break  # nothing sequenced, break out and return None
-                else:
-                    return min(sequence_dates)
 
-        return None  # if nothing is found just return None
+        step_names_udfs = MASTER_STEPS_UDFS['sequenced_step']
+        sequenced_dates = []
+
+        for process_type in step_names_udfs:
+            artifacts = self.get_artifacts(process_type=process_type, samplelimsid=lims_id)
+
+            for artifact in artifacts:
+                udf_key = step_names_udfs[process_type]
+                if artifact.parent_process and artifact.parent_process.udf.get(udf_key):
+                    sequenced_dates.append((artifact.parent_process.date_run,
+                                            artifact.parent_process.udf.get(udf_key)))
+                if artifact.parent_process and process_type == 'AUTOMATED - NovaSeq Run':
+                    sequenced_dates.append((artifact.parent_process.date_run,
+                                            artifact.parent_process.date_run))
+
+        sorted_dates = sorted(sequenced_dates, key=lambda x: x[0])
+
+        if sorted_dates:
+            log.warning("multiple sequence artifacts found for: %s", lims_id)
+
+        return sorted_dates[-1][1] if sorted_dates else None
```

Comment:
is this the same as `return min(sequence_dates)` that was the previous behavior?

Reply:
@emiliaol As we just discussed, we should take the sequence date with the most recent date_run, not the first sequqnce date, correct?

Reply:
yes